IndexFollowingIT.testUpdateRemoteConfigsDuringFollowing fails on master

Log: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob+fast+part2/4093/console
Build Scans: https://gradle-enterprise.elastic.co/s/tp2q6hwdkzcom

REPRODUCE WITH: ./gradlew ':x-pack:plugin:ccr:internalClusterTest' --tests "org.elasticsearch.xpack.ccr.IndexFollowingIT.testUpdateRemoteConfigsDuringFollowing" \
  -Dtests.seed=6CB3F4DF8251DFB5 \
  -Dtests.security.manager=true \
  -Dtests.locale=es-AR \
  -Dtests.timezone=Asia/Aden \
  -Dcompiler.java=13

Doesn't reproduce for me.
No other failures of this test for this year.

Stack trace:
```
java.lang.AssertionError: incorrect global checkpoint {"remote_cluster":"leader_cluster","follow_shard_index":"index2","follow_shard_index_uuid":"56HGhnidRZOl-eRO5HtDUw","follow_shard_shard":0,"leader_shard_index":"index1","leader_shard_index_uuid":"mlnxKU3BSPqE8AGe9MQt9A","leader_shard_shard":0,"max_read_request_operation_count":5120,"max_write_request_operation_count":5120,"max_outstanding_read_requests":12,"max_outstanding_write_requests":9,"max_read_request_size":"32mb","max_write_request_size":"9223372036854775807b","max_write_buffer_count":2147483647,"max_write_buffer_size":"512mb","max_retry_delay":"10ms","read_poll_timeout":"10ms","headers":{}}
Expected: <229L>
     but: was <-1L>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.elasticsearch.xpack.ccr.IndexFollowingIT.lambda$assertTask$63(IndexFollowingIT.java:1476)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:881)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:854)
	at org.elasticsearch.xpack.ccr.IndexFollowingIT.testUpdateRemoteConfigsDuringFollowing(IndexFollowingIT.java:1348)
```

@dnhatn by any chance this was not backported to 6.8 ? 
there was a very similar failure in that test on that branch
```
java.lang.AssertionError: incorrect global checkpoint {"remote_cluster":"leader_cluster","follow_shard_index":"index2","follow_shard_index_uuid":"6PI3qVcLS12o7i-_cGFeEw","follow_shard_shard":1,"leader_shard_index":"index1","leader_shard_index_uuid":"OiYzpvIbThCOK_FB8uuSCA","leader_shard_shard":1,"max_read_request_operation_count":9016,"max_write_request_operation_count":5120,"max_outstanding_read_requests":12,"max_outstanding_write_requests":9,"max_read_request_size":"30573350b","max_write_request_size":"9223372036854775807b","max_write_buffer_count":2147483647,"max_write_buffer_size":"512mb","max_retry_delay":"10ms","read_poll_timeout":"10ms","headers":{}}
Expected: <171L>
     but: was <-1L>
	at __randomizedtesting.SeedInfo.seed([3C886A4EEAD67EB9:FA33AFE3CB28BE11]:0)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.elasticsearch.xpack.ccr.IndexFollowingIT.lambda$assertTask$53(IndexFollowingIT.java:1386)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:906)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:880)
	at org.elasticsearch.xpack.ccr.IndexFollowingIT.testUpdateRemoteConfigsDuringFollowing(IndexFollowingIT.java:1257)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
```
https://gradle-enterprise.elastic.co/s/ysedpzc6m6ilc

--- a/server/src/main/java/org/elasticsearch/transport/RemoteClusterConnection.java
+++ b/server/src/main/java/org/elasticsearch/transport/RemoteClusterConnection.java
@@ -47,6 +47,7 @@ import org.elasticsearch.common.transport.TransportAddress;
 import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.common.util.CancellableThreads;
 import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
 import org.elasticsearch.common.util.concurrent.ThreadContext;
 import org.elasticsearch.core.internal.io.IOUtils;
 import org.elasticsearch.threadpool.ThreadPool;
@@ -65,7 +66,6 @@ import java.util.Set;
 import java.util.concurrent.ArrayBlockingQueue;
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.ExecutorService;
-import java.util.concurrent.RejectedExecutionException;
 import java.util.concurrent.Semaphore;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Consumer;
@@ -442,7 +442,7 @@ public final class RemoteClusterConnection implements TransportConnectionListene
                 ContextPreservingActionListener.wrapPreservingContext(connectListener, threadPool.getThreadContext());
             synchronized (queue) {
                 if (listener != null && queue.offer(listener) == false) {
-                    listener.onFailure(new RejectedExecutionException("connect queue is full"));
+                    listener.onFailure(new EsRejectedExecutionException("connect queue is full"));
                     return;
                 }
                 if (forceRun == false && queue.isEmpty()) {

--- a/server/src/test/java/org/elasticsearch/transport/RemoteClusterConnectionTests.java
+++ b/server/src/test/java/org/elasticsearch/transport/RemoteClusterConnectionTests.java
@@ -49,6 +49,7 @@ import org.elasticsearch.common.transport.BoundTransportAddress;
 import org.elasticsearch.common.transport.TransportAddress;
 import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.common.util.CancellableThreads;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
 import org.elasticsearch.common.util.concurrent.ThreadContext;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentFactory;
@@ -97,6 +98,7 @@ import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.iterableWithSize;
 import static org.hamcrest.Matchers.not;
 import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
 import static org.hamcrest.Matchers.sameInstance;
 import static org.hamcrest.Matchers.startsWith;
 
@@ -1199,6 +1201,67 @@ public class RemoteClusterConnectionTests extends ESTestCase {
                     assertTrue(connectionManager.nodeConnected(seedNode));
                     assertTrue(connectionManager.nodeConnected(discoverableNode));
                     assertTrue(connection.assertNoRunningConnections());
+                }
+            }
+        }
+    }
+
+    public void testPendingConnectListeners() throws IOException, InterruptedException {
+        List<DiscoveryNode> knownNodes = new CopyOnWriteArrayList<>();
+        try (MockTransportService seedTransport = startTransport("seed_node", knownNodes, Version.CURRENT);
+             MockTransportService discoverableTransport = startTransport("discoverable_node", knownNodes, Version.CURRENT)) {
+            DiscoveryNode seedNode = seedTransport.getLocalDiscoNode();
+            knownNodes.add(seedTransport.getLocalDiscoNode());
+            knownNodes.add(discoverableTransport.getLocalDiscoNode());
+            Collections.shuffle(knownNodes, random());
+            try (MockTransportService service = MockTransportService.createNewService(Settings.EMPTY, Version.CURRENT, threadPool, null)) {
+                service.start();
+                service.acceptIncomingRequests();
+                final Settings settings = Settings.builder()
+                    .put(RemoteClusterConnection.REMOTE_MAX_PENDING_CONNECTION_LISTENERS.getKey(), 1).build();
+                try (RemoteClusterConnection connection = new RemoteClusterConnection(settings, "test-cluster",
+                    seedNodes(seedNode), service, Integer.MAX_VALUE, n -> true, null, profile)) {
+                    ConnectionManager connectionManager = connection.getConnectionManager();
+                    CountDownLatch connectionOpenedLatch = new CountDownLatch(1);
+                    CountDownLatch connectionBlockedLatch = new CountDownLatch(1);
+                    connectionManager.addListener(new TransportConnectionListener() {
+                        @Override
+                        public void onConnectionOpened(Transport.Connection connection) {
+                            connectionOpenedLatch.countDown();
+                            try {
+                                connectionBlockedLatch.await();
+                            } catch (InterruptedException e) {
+                                throw new AssertionError(e);
+                            }
+                        }
+                    });
+
+                    Thread thread = new Thread(() -> connection.ensureConnected(ActionListener.wrap(() -> {})));
+                    thread.start();
+                    connectionOpenedLatch.await();
+                    connection.ensureConnected(ActionListener.wrap(() -> {}));
+                    try {
+                        int pendingConnections = randomIntBetween(1, 5);
+                        for (int i = 0; i < pendingConnections; i++) {
+                            AtomicReference<Exception> error = new AtomicReference<>();
+                            connection.ensureConnected(new ActionListener<Void>() {
+                                @Override
+                                public void onResponse(Void aVoid) {
+
+                                }
+
+                                @Override
+                                public void onFailure(Exception e) {
+                                    error.set(e);
+                                }
+                            });
+                            assertThat(error.get(), not(nullValue()));
+                            assertThat(error.get(), instanceOf(EsRejectedExecutionException.class));
+                        }
+                    } finally {
+                        connectionBlockedLatch.countDown();
+                        thread.join();
+                    }
                 }
             }
         }

